third_party/tonic/dart_weak_persistent_value.cc

```diff
@@ -0,0 +1,79 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "tonic/dart_weak_persistent_value.h"
+
+#include "tonic/dart_state.h"
+#include "tonic/scopes/dart_isolate_scope.h"
+
+namespace tonic {
+
+DartWeakPersistentValue::DartWeakPersistentValue() : value_(nullptr) {}
+
+DartWeakPersistentValue::DartWeakPersistentValue(
+    DartWeakPersistentValue&& other)
+    : dart_state_(other.dart_state_), value_(other.value_) {
+  other.dart_state_.reset();
+  other.value_ = nullptr;
+}
+
+DartWeakPersistentValue::DartWeakPersistentValue(
+    DartState* dart_state,
+    Dart_Handle object,
+    void* peer,
+    intptr_t external_allocation_size,
+    Dart_HandleFinalizer callback)
+    : value_(nullptr) {
+  Set(dart_state, object, peer, external_allocation_size, callback);
+}
+
+DartWeakPersistentValue::~DartWeakPersistentValue() {
+  Clear();
+}
+
+void DartWeakPersistentValue::Set(DartState* dart_state,
+                                  Dart_Handle object,
+                                  void* peer,
+                                  intptr_t external_allocation_size,
+                                  Dart_HandleFinalizer callback) {
+  TONIC_DCHECK(is_empty());
+  dart_state_ = dart_state->GetWeakPtr();
+  value_ = Dart_NewWeakPersistentHandle(object, peer, external_allocation_size,
+                                        callback);
+}
+
+void DartWeakPersistentValue::Clear() {
+  if (!value_) {
+    return;
+  }
+
+  auto dart_state = dart_state_.lock();
+  if (!dart_state) {
+    return;
+  }
+
+  if (Dart_CurrentIsolateGroup()) {
+    Dart_DeleteWeakPersistentHandle(value_);
+  } else if (!dart_state->IsShuttingDown()) {
+    DartIsolateScope scope(dart_state->isolate());
+    Dart_DeleteWeakPersistentHandle(value_);
+  }
+  // If it's shutting down, the handle will be deleted already.
+
+  dart_state_.reset();
+  value_ = nullptr;
+}
+
+Dart_Handle DartWeakPersistentValue::Get() {
+  if (!value_)
```

Comment:
Should we check / assert that we're not shutting down?

Reply:
Done